### PR TITLE
Vet Clinic. Milestone: Query And Update Animals Table

### PR DESCRIPTION
**Pull request by CaptainBawa**
### On this branch:

 - [ ] I inserted new data.
 - [ ] Inside a transaction:
      - Update the animals table by setting the species column to digimon for all animals that have a name ending in mon.
      - Update the animals table by setting the species column to pokemon for all animals that don't have species already set.
      - Verify that changes were made.
      - Commit the transaction.
      - Verify that changes persist after commit.
 
### Screenshot.

update-species

- [ ] Now, take a deep breath and... Inside a transaction delete all records in the animals table, then roll back the transaction.
     - After the rollback verify if all records in the animals table still exists. After that, you can start breathing as usual ;)
   
### Screenshot.

delete-rollback

- [ ] Inside a transaction:
     - Delete all animals born after Jan 1st, 2022.
     - Create a savepoint for the transaction.
     - Update all animals' weight to be their weight multiplied by -1.
     - Rollback to the savepoint
     - Update all animals' weights that are negative to be their weight multiplied by -1.
     - Commit transaction

### Screenshot.

savepoint

- [ ] Write queries to answer the following questions:
     - How many animals are there?
     - How many animals have never tried to escape?
     - What is the average weight of animals?
     - Who escapes the most, neutered or not neutered animals?
     - What is the minimum and maximum weight of each type of animal?
     - What is the average number of escape attempts per animal type of those born between 1990 and 2000?
      
### Screenshot.

math

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -24,3 +24,70 @@ SELECT * FROM animals WHERE name <> 'Gabumon';
...
 -- (including the animals with the weights that equals precisely 10.4kg or 17.3kg)
 SELECT * FROM animals WHERE weight_kg >= 10.4 AND weight_kg <= 17.3;
+
+-- New
+-- Update the animals table by setting the species column to digimon for all animals that have a name ending in mon.
+UPDATE animals SET species = 'digimon' WHERE name LIKE '%mon';
+
+-- Update the animals table by setting the species column to pokemon for all animals that don't have 
+-- species already set.
+UPDATE animals SET species = 'pokemon' WHERE species IS NULL;
```
Comment (laiifuu) [inline queries.sql:52]: Nice job with these updates :clap: However, as stated in the [project requirements](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=Inside%20a%20transaction%3A,changes%20persist%20after%20commit.), these updates should be done inside a transaction, and you should verify the changes before and after the commit by making a select query on the table.

File: queries.sql [hunk 2]
```diff
@@ -24,3 +24,70 @@ SELECT * FROM animals WHERE name <> 'Gabumon';
...
+SELECT COUNT(*) FROM animals;
+-- How many animals have never tried to escape?
+SELECT COUNT(*) FROM animals WHERE escape_attempts = 0;
+-- What is the average weight of animals?
+SELECT AVG(weight_kg) FROM animals;
+-- Who escapes the most, neutered or not neutered animals?
+SELECT neutered, COUNT(*) FROM animals WHERE escape_attempts > 0 GROUP BY neutered;
+-- What is the minimum and maximum weight of each type of animal?
+SELECT MIN(weight_kg) FROM animals;
+SELECT MAX(weight_kg) FROM animals;
```
Comment (laiifuu) [inline queries.sql]: For this question, you should use one query to find the answer, not two. It's possible to utilize `GROUP BY` for this.

File: queries.sql [hunk 3]
```diff
@@ -24,3 +24,70 @@ SELECT * FROM animals WHERE name <> 'Gabumon';
...
+-- What is the average number of escape attempts per animal type of those 
+-- born between 1990 and 2000?
+SELECT AVG(escape_attempts) FROM animals WHERE date_of_birth >= '1990-01-01' AND date_of_birth <= '2000-12-31';
```
Comment (laiifuu) [inline queries.sql]: Good work on this question :+1: However, if you read the question carefully, you'll see that it states you should find the average number `per animal type`, meaning the average number of escape attempts for the pokemons, and the average number of escape attempts for the digimons. Again, you can utilize `GROUP BY` for this.

File: schema.sql [hunk 4]
```diff
@@ -6,5 +6,8 @@ CREATE TABLE animals (
     date_of_birth DATE,
     escape_attempts INT,
     neutered BOOLEAN DEFAULT FALSE,
-    weight_kg NUMERIC
+    weight_kg NUMERIC,
```
Comment (laiifuu) [inline schema.sql]: When executing this, it gave me an error: 
Screenshot from 2023-06-14 00-00-02
I suggest you remove the `,` on line 9.

**Review by laiifuu — CHANGES_REQUESTED** [review]
# Status: CHANGES REQUIRED :recycle: 

## Hi @CaptainBawa :wave: 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
so-close



## Highlights
- [x] Organized files :heavy_check_mark: 

## Required Changes ♻️: 

- [ ] If you look back on the [project requirements](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=Inside%20a%20transaction%20update%20the%20animals%20table%20by%20setting%20the%20species%20column%20to%20unspecified.%20Verify%20that%20change%20was%20made.%20Then%20roll%20back%20the%20change%20and%20verify%20that%20the%20species%20columns%20went%20back%20to%20the%20state%20before%20the%20transaction.), however you the one which states the following: 
Screenshot from 2023-06-13 23-45-00
Kindly add the transaction for it. 

_Check the comments under the review :point_down:_

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by CaptainBawa** [comment]
### Changes have been made:

- [ ] For this question, you should use one query to find the answer, not two. It's possible to utilize GROUP BY for this.

### Screenshot.

min-max

- [ ] Good work on this question 👍 However, if you read the question carefully, you'll see that it states you should find the average number per animal type, meaning the average number of escape attempts for the pokemons, and the average number of escape attempts for the digimons. Again, you can utilize GROUP BY for this.


### Screenshot. 

avg


- [ ] Nice job with these updates 👏 However, as stated in the [project requirements](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=Inside%20a%20transaction%3A,changes%20persist%20after%20commit.), these updates should be done inside a transaction, and you should verify the changes before and after the commit by making a select query on the table.

- This query is already in a transaction. You can verify it from the previous screenshot. I have now updated it in the queries.sql

**Review by gemmen29 — APPROVED** [review]
Hi @CaptainBawa, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**